design/20220720-per-certificate-owner-ref.md

```diff
@@ -0,0 +1,62 @@
+# Design: Per-Certificate Secret Owner Reference
+
+> 🌟 This design document was written by Maël Valais on 20 July 2022 in order to facilitate Denis Romanenko's feature request presented in [#5158](https://github.com/cert-manager/cert-manager/pull/5158).
+
+cert-manager has the ability to set the owner reference field in generated Secret resources. The option is global, and takes the form of the flag `--enable-certificate-owner-ref` set in the cert-manager controller Deployment resource.
+
+Let us take an example of Certificate resource:
+
+```yaml
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: cert-1
+  namespace: ns-1
+  uid: 1e0adf8
+spec:
+  secretRef: cert-1
+```
+
+When `--enable-certificate-owner-ref` is passed to the cert-manager controller, when issuing the X.509 certificate, cert-manager will create a Secret resource that looks like this:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cert-1
+  namespace: ns-1
+  ownerReferences:
+    - controller: true
+      blockOwnerDeletion: false
+      uid: 1e0adf8
+      name: cert-1
+      kind: Certificate
+      apiVersion: cert-manager.io/v1
+data:
+  tls.crt: "..."
+  tls.key: "..."
+  ca.crt: "..."
+```
+
+The proposition is to add a new field `certificateOwnerRef` to the Certificate resource:
+
+```yaml
+apiVersion: cert-manager.io/v1
+kind: Certificate
+spec:
+  secretRef: cert-1
+  certificateOwnerRef: true  # ✨
+```
+
+It has three possible values:
+
+1. When "empty", the behavior will default to not creating an owner reference on the Secret resource, unless `--enable-certificate-owner-ref` is passed.
+2. When `true`, the default behavior as described in the "empty" case is overridden and the owner reference is always created on the Secret resource.
+3. When `false`, the default behavior as described in the "empty" case is overridden and the owner reference is never created on the Secret resource.
+
+> **⁉️ Question:** the field name `certificateOwnerRef` does not reflect the behavior that it aims to enable. A more appropriate, less confusing name could be found, e.g., `deleteSecretUponDeletion`.
```

Comment:
As @munnerz suggested in https://github.com/cert-manager/cert-manager/pull/5158#issuecomment-1189056913, the current boolean field `certificateOwnerRef` poses two questions:

1. Do we think it makes sense to have an "open string" because we think that there is a chance that we will need to extend this in the future?
2. Is the name good enough to convey the behavior of that field?

For context, James suggests:

> Something more like:
> ```yaml
> cleanupPolicy: [Delete|Never]
> ```
> may feel more natural and be more informative to our users, as well as potentially allowing us to extend the different policies we offer in future too.

Reply:
I'd also call out the Kubernetes API conventions doc which does note that bool typed fields should be avoided generally, in favour of this sort of enum value 😊

Reply:
Ah, I wasn't aware of this Kubernetes API convention, and now that you mention it, I realize that I never met any boolean in the Kubernetes API. Good catch.

I am in favor of following James's advice on this. I will update the design document to reflect this, and also mention whether or not a default value should be set when no value is given. I guess we should not default anything.

Reply:
The default should match today's behaviour (Never) - we can set this via openapi I think 😊

Reply:
We cannot set default value because this field have 3 states, "set ownerRef field", "do not set ownerRef field" and "this field do not have value and we use global setting from cert-manager"